From 0514411c4550e2c24b63e0f153057e0602c8b417 Mon Sep 17 00:00:00 2001
From: Andrei Warkentin <andrey.warkentin@gmail.com>
Date: Sat, 15 Oct 2011 02:21:45 -0400
Subject: [PATCH 2/5] Loop: Useful I/O routines shared by parsers.

This is basic functionality otherwise duplicated, like
reading/writing a kernel buffer or struct page. Provides
a generic sparse I/O routine that can be used by most
parsers (QCOW, VHD, VMDK...).

Signed-off-by: Andrei Warkentin <andrey.warkentin@gmail.com>
---
 drivers/block/Kconfig    |    4 +
 drivers/block/Makefile   |    1 +
 drivers/block/looputil.c |  322 ++++++++++++++++++++++++++++++++++++++++++++++
 include/linux/looputil.h |   82 ++++++++++++
 4 files changed, 409 insertions(+), 0 deletions(-)
 create mode 100644 drivers/block/looputil.c
 create mode 100644 include/linux/looputil.h

diff --git a/drivers/block/Kconfig b/drivers/block/Kconfig
index 6f07ec1..326fd90 100644
--- a/drivers/block/Kconfig
+++ b/drivers/block/Kconfig
@@ -286,6 +286,10 @@ config BLK_DEV_CRYPTOLOOP
 	  instead, which can be configured to be on-disk compatible with the
 	  cryptoloop device.
 
+config BLK_DEV_LOOP_UTIL
+	bool
+	depends on BLK_DEV_LOOP
+
 source "drivers/block/drbd/Kconfig"
 
 config BLK_DEV_NBD
diff --git a/drivers/block/Makefile b/drivers/block/Makefile
index 76646e9..eb3df97 100644
--- a/drivers/block/Makefile
+++ b/drivers/block/Makefile
@@ -28,6 +28,7 @@ obj-$(CONFIG_BLK_DEV_OSD)	+= osdblk.o
 obj-$(CONFIG_BLK_DEV_UMEM)	+= umem.o
 obj-$(CONFIG_BLK_DEV_NBD)	+= nbd.o
 obj-$(CONFIG_BLK_DEV_CRYPTOLOOP) += cryptoloop.o
+obj-$(CONFIG_BLK_DEV_LOOP_UTIL)	+= looputil.o
 obj-$(CONFIG_VIRTIO_BLK)	+= virtio_blk.o
 
 obj-$(CONFIG_VIODASD)		+= viodasd.o
diff --git a/drivers/block/looputil.c b/drivers/block/looputil.c
new file mode 100644
index 0000000..221cbe1
--- /dev/null
+++ b/drivers/block/looputil.c
@@ -0,0 +1,322 @@
+/*
+   Useful routines shared by loop parsers.
+
+   Copyright (C) 2011 Andrei Warkentin <andreiw@vmware.com>
+
+   This module is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 2 of the License, or
+   (at your option) any later version.
+
+   This module is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this module; if not, write to the Free Software
+   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#include <linux/loop.h>
+#include <linux/looputil.h>
+
+static int loop_kbuf_to_bio(struct bio *bio,
+			    u8 *kbuf,
+			    size_t len)
+{
+	unsigned int bv_len;
+	unsigned int bv_offset;
+	struct bio_vec *bvec;
+
+	/*
+	 * Fake a BIO containing bvec's describing a physically
+	 * contiguous kernel buffer. bio_add_page is too smart
+	 * and relies on a present rq. In our case, we just want
+	 * to use the same file I/O logic, which just happens to
+	 * use BIO. It might be nicer to redo everything in loop-ng
+	 * to use sg's and convert BIO nvecs to sg inside
+	 * do_bio_filebacked.
+	 *
+	 */
+	while (len) {
+		bv_offset = ((uintptr_t) kbuf) & ~PAGE_MASK;
+		bv_len = min(len, (size_t) PAGE_SIZE - bv_offset);
+		bvec = &bio->bi_io_vec[bio->bi_vcnt];
+		bvec->bv_page = virt_to_page(kbuf);
+		bvec->bv_len = bv_len;
+		bvec->bv_offset = bv_offset;
+		bio->bi_vcnt++;
+		len -= bv_len;
+		kbuf += bv_len;
+	}
+	return 0;
+}
+
+static int transfer_zero(struct loop_device *lo, int cmd,
+			 struct page *raw_page, unsigned raw_off,
+			 struct page *loop_page, unsigned loop_off,
+			 int size, sector_t real_block)
+{
+	char *buf;
+
+	if (cmd == READ)
+		buf = kmap_atomic(loop_page, KM_USER0) + loop_off;
+	else
+		buf = kmap_atomic(raw_page, KM_USER0) + raw_off;
+
+	memset(buf, 0, size);
+	kunmap_atomic(buf, KM_USER0);
+	cond_resched();
+	return 0;
+}
+
+int loop_write_zero(struct loop_device *lo,
+		    struct lo_file *lo_file,
+		    size_t len,
+		    loff_t pos)
+{
+	unsigned vecs;
+	size_t cur_len;
+	int ret = 0;
+	struct bio *bio = NULL;
+
+	while (len) {
+		cur_len = min(len, (size_t) BIO_MAX_SIZE);
+		vecs = (cur_len + PAGE_SIZE - 1) >> PAGE_SHIFT;
+		bio = bio_alloc(GFP_KERNEL, vecs);
+		BUG_ON(!bio);
+
+		/*
+		 * We end up describing a bogus range, which is safe,
+		 * since it is ignored by transfer_zero.
+		 */
+		ret = loop_kbuf_to_bio(bio, NULL, cur_len);
+		if (ret)
+			goto out;
+
+		ret = loop_send(lo, lo_file, bio, pos, transfer_zero);
+		if (ret)
+			goto out;
+
+		bio_put(bio);
+		len -= cur_len;
+		pos += cur_len;
+	}
+
+out:
+	if (len)
+		bio_put(bio);
+	return ret;
+}
+
+int loop_write_kbuf(struct loop_device *lo,
+		    struct lo_file *lo_file,
+		    void *kbuf,
+		    size_t len,
+		    loff_t pos)
+{
+	unsigned vecs;
+	size_t cur_len;
+	int ret = 0;
+	u8 *buf = kbuf;
+	struct bio *bio = NULL;
+
+	while (len) {
+		cur_len = min(len, (size_t) BIO_MAX_SIZE);
+		vecs = (cur_len + PAGE_SIZE - 1) >> PAGE_SHIFT;
+		bio = bio_alloc(GFP_KERNEL, vecs);
+		BUG_ON(!bio);
+
+		/*
+		 * We end up describing a bogus range, which is safe,
+		 * since it is ignored by transfer_zero.
+		 */
+		ret = loop_kbuf_to_bio(bio, buf, cur_len);
+		if (ret)
+			goto out;
+
+		ret = loop_send(lo, lo_file, bio, pos, NULL);
+		if (ret)
+			goto out;
+
+		bio_put(bio);
+		len -= cur_len;
+		pos += cur_len;
+		buf += cur_len;
+	}
+
+out:
+	if (len)
+		bio_put(bio);
+	return ret;
+}
+
+int loop_read_kbuf(struct loop_device *lo,
+		   struct lo_file *lo_file,
+		   void *kbuf,
+		   size_t len,
+		   loff_t pos)
+{
+	unsigned vecs;
+	size_t cur_len;
+	int ret = 0;
+	u8 *buf = kbuf;
+	struct bio *bio = NULL;
+
+	while (len) {
+		cur_len = min(len, (size_t) BIO_MAX_SIZE);
+		vecs = (cur_len + PAGE_SIZE - 1) >> PAGE_SHIFT;
+		bio = bio_alloc(GFP_KERNEL, vecs);
+		BUG_ON(!bio);
+
+		/*
+		 * We end up describing a bogus range, which is safe,
+		 * since it is ignored by transfer_zero.
+		 */
+		ret = loop_kbuf_to_bio(bio, buf, cur_len);
+		if (ret)
+			goto out;
+
+		ret = loop_recv(lo, lo_file, bio, pos, NULL);
+		if (ret)
+			goto out;
+
+		bio_put(bio);
+		len -= cur_len;
+		pos += cur_len;
+		buf += cur_len;
+	}
+
+out:
+	if (len)
+		bio_put(bio);
+	return ret;
+}
+
+int loop_read_page(struct loop_device *lo,
+		   struct lo_file *lo_file,
+		   struct page *page,
+		   unsigned int offset,
+		   unsigned int len,
+		   loff_t pos)
+{
+	int ret;
+	struct bio *bio = bio_alloc(GFP_KERNEL, 1);
+	BUG_ON(!bio);
+
+	bio->bi_io_vec->bv_page = page;
+	bio->bi_io_vec->bv_offset = offset;
+	bio->bi_io_vec->bv_len = len;
+	bio->bi_vcnt = 1;
+	ret = loop_recv(lo, lo_file, bio, pos, NULL);
+	bio_put(bio);
+	return ret;
+}
+
+int loop_write_page(struct loop_device *lo,
+		    struct lo_file *lo_file,
+		    struct page *page,
+		    unsigned int offset,
+		    unsigned int len,
+		    loff_t pos)
+{
+	int ret;
+	struct bio *bio = bio_alloc(GFP_KERNEL, 1);
+	BUG_ON(!bio);
+
+	bio->bi_io_vec->bv_page = page;
+	bio->bi_io_vec->bv_offset = offset;
+	bio->bi_io_vec->bv_len = len;
+	bio->bi_vcnt = 1;
+	ret = loop_send(lo, lo_file, bio, pos, NULL);
+	bio_put(bio);
+	return ret;
+}
+
+static int loop_sparse_bvec(struct loop_device *lo,
+			    struct lo_file *lo_file,
+			    struct bio_vec *bvec,
+			    loff_t pos,
+			    bool write,
+			    loop_sparse_to_t sparse_to)
+{
+	u8 *raw_buf;
+	struct loop_sparse_in in;
+	struct loop_sparse_out out;
+	unsigned int bv_len = bvec->bv_len;
+	int ret = 0;
+
+	in.write = write;
+	in.pos = pos;
+
+	while (bv_len) {
+		in.len = bv_len;
+		ret = sparse_to(lo, &in, &out);
+
+		switch (ret) {
+		case LOOP_SPARSE_ZERO:
+		case LOOP_SPARSE_MEM:
+
+			/*
+			 * These are used for sparse empty and
+			 * compressed sections only.
+			 */
+			BUG_ON(write);
+			raw_buf = kmap_atomic(bvec->bv_page, KM_USER0) +
+				bvec->bv_offset + bvec->bv_len - bv_len;
+			if (ret == LOOP_SPARSE_ZERO)
+				memset(raw_buf, 0, out.len);
+			else
+				memcpy(raw_buf, out.offset, out.len);
+			kunmap_atomic(raw_buf, KM_USER0);
+			ret = 0;
+			break;
+		case LOOP_SPARSE_IO:
+			if (write)
+				ret = loop_write_page(lo, out.lo_file,
+						      bvec->bv_page,
+						      bvec->bv_offset + bvec->bv_len -
+						      bv_len, out.len, out.pos);
+			else
+				ret = loop_read_page(lo, out.lo_file,
+						     bvec->bv_page,
+						     bvec->bv_offset + bvec->bv_len -
+						     bv_len, out.len, out.pos);
+			break;
+		}
+
+		if (ret) {
+			printk(KERN_ERR "%s: failed to %s: %d\n",
+			       lo->lo_disk->disk_name,
+                               write ? "write" : "read",
+			       ret);
+			break;
+		}
+
+		bv_len -= out.len;
+		in.pos += out.len;
+	}
+
+	return ret;
+}
+
+int loop_sparse_io(struct loop_device *lo,
+		   struct lo_file *lo_file,
+		   struct bio *bio,
+		   loop_sparse_to_t sparse_to)
+{
+	struct bio_vec *bvec;
+	int i = 0;
+	int ret = 0;
+	loff_t pos = ((loff_t) bio->bi_sector << 9);
+
+	bio_for_each_segment(bvec, bio, i) {
+		ret = loop_sparse_bvec(lo, lo_file, bvec, pos,
+				       bio_rw(bio) == WRITE,
+				       sparse_to);
+		pos += bvec->bv_len;
+	}
+
+	return ret;
+}
diff --git a/include/linux/looputil.h b/include/linux/looputil.h
new file mode 100644
index 0000000..f6b972d
--- /dev/null
+++ b/include/linux/looputil.h
@@ -0,0 +1,82 @@
+/*
+  Useful routines for loop parsers.
+
+  Copyright (C) 2011 Andrei Warkentin <andreiw@vmware.com>
+
+  This module is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2 of the License, or
+  (at your option) any later version.
+
+  This module is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this module; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+*/
+
+#ifndef _LINUX_LOOP_UTIL_H
+#define _LINUX_LOOP_UTIL_H
+
+#include <linux/loop.h>
+
+int loop_write_zero(struct loop_device *lo,
+		    struct lo_file *lo_file,
+		    size_t len,
+		    loff_t pos);
+int loop_write_kbuf(struct loop_device *lo,
+		    struct lo_file *lo_file,
+		    void *kbuf,
+		    size_t len,
+		    loff_t pos);
+int loop_read_kbuf(struct loop_device *lo,
+		   struct lo_file *lo_file,
+		   void *kbuf,
+		   size_t len,
+		   loff_t pos);
+int loop_read_page(struct loop_device *lo,
+		   struct lo_file *lo_file,
+		   struct page *page,
+		   unsigned int offset,
+		   unsigned int len,
+		   loff_t pos);
+int loop_write_page(struct loop_device *lo,
+		    struct lo_file *lo_file,
+		    struct page *page,
+		    unsigned int offset,
+		    unsigned int len,
+		    loff_t pos);
+
+enum {
+	LOOP_SPARSE_IO,
+	LOOP_SPARSE_MEM,
+	LOOP_SPARSE_ZERO,
+};
+
+struct loop_sparse_in {
+	loff_t pos;
+	unsigned int len;
+	bool write;
+};
+
+struct loop_sparse_out {
+	union {
+		loff_t pos;
+		void  *offset;
+	};
+	struct lo_file *lo_file;
+	unsigned int len;
+};
+
+typedef int (*loop_sparse_to_t)(struct loop_device *lo,
+				struct loop_sparse_in *in,
+				struct loop_sparse_out *out);
+int loop_sparse_io(struct loop_device *lo,
+		   struct lo_file *lo_file,
+		   struct bio *bio,
+		   loop_sparse_to_t sparse_to);
+
+#endif /* _LINUX_LOOP_UTIL_H */
-- 
1.7.7

